provider/oci/environ.go

```diff
@@ -223,42 +378,313 @@ func (e *Environ) Destroy(ctx envcontext.ProviderCallContext) error {
+	if apiPort == 0 {
+		return cloudcfg, nil
+	}
+
+	operatingSystem, err := jujuseries.GetOSFromSeries(series)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	switch operatingSystem {
+	case os.Ubuntu:
+		fwCmd := fmt.Sprintf(
+			"/sbin/iptables -I INPUT -p tcp --dport %d -j ACCEPT", apiPort)
+		cloudcfg.AddRunCmd(fwCmd)
+		cloudcfg.AddScripts("/etc/init.d/netfilter-persistent save")
+	case os.CentOS:
+		fwCmd := fmt.Sprintf("firewall-cmd --zone=public --add-port=%d/tcp --permanent", apiPort)
+		cloudcfg.AddRunCmd(fwCmd)
+	}
+	return cloudcfg, nil
 }
 
 // StartInstance implements environs.InstanceBroker.
 func (e *Environ) StartInstance(ctx envcontext.ProviderCallContext, args environs.StartInstanceParams) (*environs.StartInstanceResult, error) {
-	return nil, errors.NotImplementedf("StartInstance")
+	// var types []instances.InstanceType
+
+	if args.ControllerUUID == "" {
+		return nil, errors.NotFoundf("Controller UUID")
+	}
+
+	networks, err := e.ensureNetworksAndSubnets(ctx, args.ControllerUUID, e.Config().UUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	zones, err := e.AvailabilityZones(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	zone := zones[0].Name()
+	network := networks[zone][0]
+	// refresh the global image cache
+	// this only hits the API every 30 minutes, otherwise just retrieves
+	// from cache
+	imgCache, err := refreshImageCache(e.Compute, e.ecfg().compartmentID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Debugf("Image cache contains: %v", imgCache)
+	series := args.Tools.OneSeries()
+	arches := args.Tools.Arches()
+
+	types := imgCache.SupportedShapes(series)
+
+	defaultType := string(VirtualMachine)
+	if args.Constraints.VirtType == nil {
+		args.Constraints.VirtType = &defaultType
+	}
+
+	// check if we find an image that is compliant with the
+	// constraints provided in the oracle cloud account
+	args.ImageMetadata = imgCache.ImageMetadata(series, *args.Constraints.VirtType)
+
+	spec, image, err := findInstanceSpec(
+		args.ImageMetadata,
+		types,
+		&instances.InstanceConstraint{
+			Series:      series,
+			Arches:      arches,
+			Constraints: args.Constraints,
+		},
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	tools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Tracef("agent binaries: %v", tools)
+	if err = args.InstanceConfig.SetTools(tools); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if err = instancecfg.FinishInstanceConfig(
+		args.InstanceConfig,
+		e.Config(),
+	); err != nil {
+		return nil, errors.Trace(err)
+	}
+	hostname, err := e.namespace.Hostname(args.InstanceConfig.MachineId)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	tags := args.InstanceConfig.Tags
+
+	var apiPort int
+	var desiredStatus ociCore.InstanceLifecycleStateEnum
+	// Wait for controller to actually be running
```

Comment:
can you explain this a little bit more? I'm a bit confused why launching a new instance is dependent on the controller.

Maybe it's just a different comment:
// If this is a new controller, we need to wait for it to be Running before returning.